### PR TITLE
Ensure that `self.outPath == ./.`

### LINKED ISSUE #7263: The definition of `self` is inconsistent in flakes
**Describe the bug**

When `flake.nix` is not located at the root of a `git` repository I have found a few scenarios in which the definition of `self` is inconsistent. Sometimes `self` refers to the entire repository, sometimes `self` refers to the directory that `flake.nix` resides in (and therefore excludes the rest of the repository code).

**Steps To Reproduce**

1. Clone [this](https://github.com/aanderse/flakes) simple repository and briefly examine it.
2. Follow the [README.md](https://github.com/aanderse/flakes/blob/master/README.md)

```
$ cd subdir/
$ nix repl                                                                                                                                                                                            
Welcome to Nix 2.8.1. Type :? for help.

nix-repl> let flake = builtins.getFlake (toString ./.); in builtins.readFile ("${flake.self-as-an-output}/test.txt")
copying '/home/aaron/flakes-test/subdir'"this is a text file at the deeper inside of my git repository\n"

nix-repl> :lf .#                                                                                                     
Added 12 variables.

nix-repl> builtins.readFile ("${outputs.self-as-an-output}/test.txt")                                                
"this is a text file at the root of my git repository\n"
```

Why do i get different results? :thinking:

**Expected behavior**

I would expect that `self` is consistently referring to the same path within my repository and includes all the same code, regardless of how I reference `self`.

**`nix-env --version` output**

`Nix 2.8.1`

### PR DESCRIPTION
**Pull request by roberth**
Ensure that `self.outPath == ./.` in cases where it wasn't.

# Motivation

Fix a bug.

# Context
 - Closes #7263

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

### CODE REVIEW

File: src/libexpr/flake/call-flake.nix [hunk 1]
```diff
@@ -9,14 +9,18 @@ let
...
             if key == lockFile.root
             then rootSrc
             else fetchTree (node.info or {} // removeAttrs node.locked ["dir"]);
 
           subdir = if key == lockFile.root then rootSubdir else node.locked.dir or "";
 
-          flake = import (sourceInfo + (if subdir != "" then "/" else "") + subdir + "/flake.nix");
+          outPath = rawSourceInfo + ((if subdir == "" then "" else "/") + subdir);
+
+          sourceInfo = rawSourceInfo // { inherit outPath; };
```
Comment (Ericson2314) [inline src/libexpr/flake/call-flake.nix:21]:
```suggestion
          # overrides `outPath` in `rawSourceInfo` which is repo root
          sourceInfo = rawSourceInfo // { inherit outPath; };
```

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team

- agreement that we want this
- @roberth: should add a release note
    - (some discussion of cost-benefit-ratio of adding release notes about bug fixes)
    - agreement
- @edolstra: `lastModified` is valid for the tree
    - @roberth: that is only information for `sourceInfo`, not the flake itself
-  assigned to @thufschmitt

**Review by thufschmitt — APPROVED** [review]
Looks good :+1: 

Thanks!

**Comment by NickCao** [comment]
They are still, not actually equal:
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs";
  };
  outputs = { self, nixpkgs }: with nixpkgs.legacyPackages.x86_64-linux; {
    packages.x86_64-linux = {
      good = runCommand "good" { path = toString self; } ''
        echo "$path"
        ls "$path" && echo "self is a proper path and has string context"
      '';
      bad = runCommand "bad" { path = toString ./.; } ''
        echo "$path"
        ls "$path" && echo "while ./. is not and is thus not accessible"
      '';
    };
  };
}
```

```
$ nix build .#good -L
good> /nix/store/ci2sf5rmwz8d54qk9n0v0bm7hbmlfxfk-source
good> flake.lock  flake.nix
good> self is a proper path and has string context
$ nix build .#bad -L
bad> /nix/store/ci2sf5rmwz8d54qk9n0v0bm7hbmlfxfk-source
bad> ls: cannot access '/nix/store/ci2sf5rmwz8d54qk9n0v0bm7hbmlfxfk-source': No such file or directory
```

**Comment by NickCao** [comment]
Actually more of an issue with `toString`, while string interpolation correctly works (just unnecessarily copies the flake again).

**Comment by roberth** [comment]
> Actually more of an issue with `toString`, while string interpolation correctly works (just unnecessarily copies the flake again).

I agree. See #8428

**Pull request merged**